### ISSUE TITLE
How do I reset the editor without losing the plugins?

### ISSUE THREAD

**Issue by nadimtuhin** [draft-js-plugins/draft-js-plugins#251]
How do I reset the editor without losing the plugins? For example 

```js
    reset(){
        this.setState({
            editorState: EditorState.createEmpty()
        })
    }
```

this resets the editor but loses the decorators

**Comment by nadimtuhin** [comment]
```js
    reset(){
        const empty = EditorState.createEmpty();
        const editorState = EditorState.moveFocusToEnd(empty);

        this.setState({
            editorState: EditorState.createEmpty()
        })
    }

```

this does not work either. but when I do it in a simple draft instance without plugins wrapper it works

**Comment by nadimtuhin** [comment]
its to be noted that previously I could create an empty state in this way.

**Comment by adrianmc** [comment]
You have to push onto your existing editor state. So first you must create a contentState and then use the push method from EditorState to create a new EditorState that preserves your old editorState's plugins. See this for more info: https://facebook.github.io/draft-js/docs/api-reference-editor-state.html#push

**Comment by nadimtuhin** [comment]
Thanks. Its working now.

```js

    reset = () => {
        const editorState = EditorState.push(this.state.editorState, ContentState.createFromText(''));
        this.setState({ editorState });
    };

```

**Comment by adrianmc** [comment]
@nadimtuhin Awesome!

**Comment by Stanley** [comment]
This should be documented somewhere :smile:

**Comment by nikgraf** [comment]
@Stanley document here now: https://github.com/draft-js-plugins/draft-js-plugins/blob/master/FAQ.md 😄

**Comment by Stanley** [comment]
Awesome job! Thank you :smile:

**Comment by GabiGrin** [comment]
Can't it easily be fixed by adding what happens here - https://github.com/draft-js-plugins/draft-js-plugins/blob/master/draft-js-plugins-editor/src/Editor/index.js#L58 to some other lifecycle hook that happens each time the state is renewed?
IMO replacing editor states should be a common scenario.

**Comment by hasnainshafqatmlt** [comment]
Hello guys after doing reset editor mention list working but not first time.
i've to right first few text then mention list works. if after submit message that clear the editor and then i search mention list by typing '@' the it does not work even after writing text then finding mention list cursor goes to start of the editor.
please see the attachment for better understanding what happen.
how i'm clear editor here is the code.

```
resetEditor() {
        $('#richEditor-actions').fadeOut();
        const editorState = EditorState.push(this.state.editorState, ContentState.createFromText(''));
        this.setState({ editorState });
};
```
![sample](https://user-images.githubusercontent.com/6399158/32447292-75eca564-c32d-11e7-837a-999238110f28.gif)

full code 

```
// Copyright (c) 2015 Plunk, Inc. All Rights Reserved.
// See License.txt for license information.

import React from 'react';
import InlineStyleControls from './inline_style_controls.jsx';
import BlockStyleControls from './block_style_controls.jsx';
import * as Utils from 'utils/utils.jsx';
import AppDispatcher from '../../dispatcher/app_dispatcher.jsx';
import PostStore from 'stores/post_store.jsx';
import UserStore from 'stores/user_store.jsx';
import Client from 'client/web_client.jsx';
import * as GlobalActions from 'actions/global_actions.jsx';
import ChannelStore from 'stores/channel_store.jsx';
import {stateToHTML} from 'draft-js-export-html';
import Constants from 'utils/constants.jsx';

import {EditorState, RichUtils, getDefaultKeyBinding, convertFromHTML, ContentState} from 'draft-js';
import Editor from 'draft-js-plugins-editor';
import createMentionPlugin, { defaultSuggestionsFilter } from 'draft-js-mention-plugin';
import { fromJS } from 'immutable';
import 'draft-js-mention-plugin/lib/plugin.css';


const positionSuggestions = ({ state, props }) => {
  let transform;
  let transition;

  if (state.isActive && props.suggestions.size > 0) {
    transform = 'scaleY(1)';
    transition = 'all 0.25s cubic-bezier(.3,1.2,.2,1)';
  } else if (state.isActive) {
    transform = 'scaleY(0)';
    transition = 'all 0.25s cubic-bezier(.3,1,.2,1)';
  }

  return {
    transform: 'scaleY(1)',
    transition: 'all 0.25s cubic-bezier(.3,1.2,.2,1)',
    top: 'auto',
    left: '0',
    right: '0',
    bottom: '100%',
    width: '100%',
    maxWidth: '100%',
    position: 'absolute'
  };
};

const ActionTypes = Constants.ActionTypes;


const styleMap = {
  CODE: {
    backgroundColor: 'rgba(0, 0, 0, 0.05)',
    fontFamily: '"Inconsolata", "Menlo", "Consolas", monospace',
    fontSize: 16,
    padding: 2,
  },
};
function getBlockStyle(block) {
  switch (block.getType()) {
    case 'blockquote': return 'RichEditor-blockquote';
    default: return null;
  }
}

export default class RichContent extends React.Component {

     constructor(props) {
        super(props);

        this.onEditorChange = this.onEditorChange.bind(this);
        this.getCurrentDraft = this.getCurrentDraft.bind(this);
        this.handleKeyCommand = this._handleKeyCommand.bind(this);
        this.myKeyBindingFn = this.myKeyBindingFn.bind(this);
        this.onTab = this._onTab.bind(this);
        this.toggleBlockType = this._toggleBlockType.bind(this);
        this.toggleInlineStyle = this._toggleInlineStyle.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onUpArrow = this.onUpArrow.bind(this);
        this.sendMessage = this.sendMessage.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.setDefaultState = this.setDefaultState.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);

        PostStore.clearDraftUploads();
        const draft = this.getCurrentDraft();

        //this.mentionPlugin = createMentionPlugin();
        this.mentionPlugin = createMentionPlugin({
          positionSuggestions,
          mentionPrefix: '@',
        });

        this.state = {
          editorState: EditorState.createEmpty(),
          messageText: draft.messageText,
          uploadsInProgress: draft.uploadsInProgress,
          submitting: false,
          previews: draft.previews,
          channelId: ChannelStore.getCurrentId(),
          suggestions: fromJS([]),
        };
      }

        
    componentDidMount() {
      ChannelStore.addChangeListener(this.onChange);
    }

    componentWillUnmount() {
      ChannelStore.removeChangeListener(this.onChange);
    }

     
    getCurrentDraft() {
        const draft = PostStore.getCurrentDraft();
        const safeDraft = {previews: [], messageText: '', uploadsInProgress: []};

        if (draft) {
            if (draft.message) {
                safeDraft.messageText = draft.message;
            }
            if (draft.previews) {
                safeDraft.previews = draft.previews;
            }
            if (draft.uploadsInProgress) {
                safeDraft.uploadsInProgress = draft.uploadsInProgress;
            }
        }

        return safeDraft;
    }

    onChange() {
        const channelId = ChannelStore.getCurrentId();
        if (this.state.channelId !== channelId) {
            const draft = this.getCurrentDraft();
            console.log('draft >>', draft , '***', draft.messageText)

            //this.setDefaultState();
            this.setState({
              channelId,
              messageText: draft.messageText,
              initialText: draft.messageText,
              submitting: false,
              serverError: null,
              postError: null,
              previews: draft.previews,
              uploadsInProgress: draft.uploadsInProgress
            });
        }
    }

    setDefaultState() {
      const draft = this.getCurrentDraft();
      let editorState;
      if (draft.messageText) {
        const blocksFromHTML = convertFromHTML(draft.messageText);
        const contentState = ContentState.createFromBlockArray(blocksFromHTML);
        editorState = EditorState.createWithContent(contentState);
      } else {
        editorState = EditorState.createEmpty();
      }
      this.setState({editorState});
    }

    onUpArrow() {
      const contentState = this.state.editorState.getCurrentContent();

      if (!contentState.hasText()) {
        const channelId = ChannelStore.getCurrentId();
        const lastPost = PostStore.getCurrentUsersLatestPost(channelId);
        if (!lastPost) {
            return;
        }

        let type;
        if (lastPost.root_id && lastPost.root_id.length > 0) {
            type = Utils.localizeMessage('create_post.comment', 'Comment');
        } else {
            type = Utils.localizeMessage('create_post.post', 'Post');
        }

        AppDispatcher.handleViewAction({
            type: ActionTypes.RECEIVED_EDIT_POST,
            refocusId: '#post_textbox',
            title: type,
            message: lastPost.message,
            postId: lastPost.id,
            channelId: lastPost.conversation_id,
            comments: lastPost.total_comments
        });
      }
    }

    myKeyBindingFn(e: SyntheticKeyboardEvent): string {
      console.log('e.keyCode >>>', e.keyCode)
      if (e.keyCode === 13 && !e.nativeEvent.shiftKey) {
        return 'enterEvent';
      }
      return getDefaultKeyBinding(e);
    }


    _handleKeyCommand(command, editorState) {
      console.log('_handleKeyCommand >>', command, editorState)
      if (command === 'enterEvent') {
        console.log(' save message')
        this.handleSubmit();
        return true;
      }
      const newState = RichUtils.handleKeyCommand(editorState, command);
      if (newState) {
        this.onEditorChange(newState);
        return true;
      }
      return false;
    }

    _onTab(e) {
      const maxDepth = 4;
      this.onEditorChange(RichUtils.onTab(e, this.state.editorState, maxDepth));
    }

    _toggleBlockType(blockType) {
      this.onEditorChange(
        RichUtils.toggleBlockType(
          this.state.editorState,
          blockType
        )
      );
    }

    _toggleInlineStyle(inlineStyle) {
      this.onEditorChange(
        RichUtils.toggleInlineStyle(
          this.state.editorState,
          inlineStyle
        )
      );
    }

    onEditorChange(editorState) {
          const contentState = editorState.getCurrentContent();
          const messageText = stateToHTML(this.state.editorState.getCurrentContent())
          
          if (contentState.hasText()) {
            $('#richEditor-actions').fadeIn();
          } else {
            $('#richEditor-actions').fadeOut();
          }

          this.setState({editorState, messageText});

          const draft = PostStore.getCurrentDraft();
          draft.message = messageText;
          PostStore.storeCurrentDraft(draft);
        }

      onSearchChange({ value }) {
        console.log('value >>', value)
        Client.getUsersSuggestions(value,
            (data) => {
                const mentionsList = [];
                data.forEach((user) => {
                  mentionsList.push({
                    avatar: user.picture,
                    name: user.full_name,
                    link: `@${user.username}`
                  });
                })
                this.setState({
                  //suggestions: defaultSuggestionsFilter(value, fromJS(mentionsList)),
                  suggestions: fromJS(mentionsList),
                });

            }
        );

        
      };

      onAddMention = () => {
        // get the mention object selected
      }

      focus = () => {
        this.editor.focus();
      };

      resetEditor() {
        $('#richEditor-actions').fadeOut();
        const editorState = EditorState.push(this.state.editorState, ContentState.createFromText(''));
        this.setState({ editorState });
      };


      handleSubmit() {
        const contentState = this.state.editorState.getCurrentContent();

        if (this.state.uploadsInProgress.length > 0 || this.state.submitting) {
            return;
        }

        const post = {};
        post.files = [];
        post.message = this.state.messageText;

        if (!contentState.hasText() && this.state.previews.length === 0) {
            return;
        }

        // if (post.message.length > Constants.CHARACTER_LIMIT) {
        //     this.setState({postError: `Post length must be less than ${Constants.CHARACTER_LIMIT} characters.`});
        //     return;
        // }

        this.setState({submitting: true, serverError: null});

        this.sendMessage(post);

        //this.props.showSearchBar(false);
    }

    sendMessage(post) {
        const channel = ChannelStore.get(this.state.channelId);
        const time = Utils.getTimestamp();
        const userId = UserStore.getCurrentId();
        const user = UserStore.getCurrentUser();

        post.conversation_id = this.state.channelId;
        post.channel = channel;

        post.pending_post_id = `${userId}:${time}`;
        post.created_at = Date();
        post.user_id = userId;

        post.author = {
            color: user.color,
            full_name: user.first_name + ' ' + user.last_name,
            id: user.id,
            picture: user.picture,
            username: user.username
        };

        const previews = this.state.previews;

        if (previews.length) {
            post.files = previews;
        }

        GlobalActions.emitUserPostedEvent(post);
        this.setState({messageText: '', submitting: false, postError: null, previews: [], serverError: null});

        Client.createPost(post,
            () => {
                PostStore.removePendingPost(post.conversation_id, post.pending_post_id);
                this.resetEditor();
                this.setState({
                    submitting: false
                });
            },
            (err) => {
                if (err.id === 'api.post.create_post.root_id.app_error') {
                    // this should never actually happen since you can't reply from this textbox
                    //this.showPostDeletedModal();

                    PostStore.removePendingPost(post.pending_post_id);
                } else {
                    post.state = Constants.POST_FAILED;
                    PostStore.updatePendingPost(post);
                }

                this.setState({
                    submitting: false
                });
            }
        );
    }

    render() {
        //const {editorState} = this.state;

        //  const { MentionSuggestions } = this.mentionPlugin;
        // const plugins = [this.mentionPlugin];

        //   // If the user changes block type before entering any text, we can
        //   // either style the placeholder or hide it. Let's just hide it now.
          let className = 'richEditor-editor';
          var contentState = this.state.editorState.getCurrentContent();
          if (!contentState.hasText()) {
            if (contentState.getBlockMap().first().getType() !== 'unstyled') {
              className += ' richEditor-hidePlaceholder';
            }
          }

 const { MentionSuggestions } = this.mentionPlugin;
    const plugins = [this.mentionPlugin];

        return (
          <div className='richEditor-root'>            
            <div className='pull-left col-100 editor-actions' id='richEditor-actions'>
              <div className='pull-left p-rel rich-text-size'>
                <a
                  href='#'
                  className='dropdown-toggle theme'
                  type='button'
                  id='rich-text-size'
                  data-toggle='dropdown'
                  aria-expanded='true'
                >
                  <i className='icon-size-i'></i>
                </a>
                <ul
                    className='dropdown-menu dropdown-theme left-auto'
                    role='menu'
                    aria-labelledby='rich-text-size'
                >
                  <BlockStyleControls
                    editorState={this.state.editorState}
                    onToggle={this.toggleBlockType}
                    size={true}
                  />
                </ul>
              </div>
              <BlockStyleControls
                editorState={this.state.editorState}
                onToggle={this.toggleBlockType}
                size={false}
              />
              <InlineStyleControls
                editorState={this.state.editorState}
                onToggle={this.toggleInlineStyle}
              />
            </div>  
            <div className={`${className}`} onClick={this.focus}>
              <Editor
                blockStyleFn={getBlockStyle}
                customStyleMap={styleMap}
                editorState={this.state.editorState}
                handleKeyCommand={this.handleKeyCommand}
                keyBindingFn={this.myKeyBindingFn}
                onChange={this.onEditorChange}
                onTab={this.onTab}
                onUpArrow={this.onUpArrow}
                plugins={plugins}
                ref={(element) => { this.editor = element; }}
                placeholder="Write a message..."
                spellCheck={true}
              />
              <MentionSuggestions
                onSearchChange={this.onSearchChange}
                suggestions={this.state.suggestions}
                onAddMention={this.onAddMention}
                class='mention-list'
                className='mention-list2'
              />
            </div>
          </div>
        );
    }
}


```

**Comment by hasnainshafqatmlt** [comment]
Hy guys, 
mention list does not work if i set initial state in editor 
anyone can help ?

```
setDefaultState() {
      const draft = this.getCurrentDraft();
      let editorState;
      if (draft.messageText) {
        const blocksFromHTML = convertFromHTML(draft.messageText);
        const contentState = ContentState.createFromBlockArray(blocksFromHTML);
        editorState = EditorState.createWithContent(contentState);
      } else {
        editorState = EditorState.createEmpty();
      }
      this.setState({editorState});
    }
```

**Comment by hasnainshafqatmlt** [comment]
Hy guys, 
mention list does not work if i set initial state in editor 
anyone can help ?

```
setDefaultState() {
      const draft = this.getCurrentDraft();
      let editorState;
      if (draft.messageText) {
        const blocksFromHTML = convertFromHTML(draft.messageText);
        const contentState = ContentState.createFromBlockArray(blocksFromHTML);
        editorState = EditorState.createWithContent(contentState);
      } else {
        editorState = EditorState.createEmpty();
      }
      this.setState({editorState});
    }
```

**Comment by hasnainshafqatmlt** [comment]
found the solution by replacing 
```
editorState = EditorState.createWithContent(contentState);
```
with
```
editorState = EditorState.push(this.state.editorState, contentState);
```
final code 

```
setDefaultState() {
      const draft = this.getCurrentDraft();
      let editorState;
      if (draft.messageText) {
        const blocksFromHTML = convertFromHTML(draft.messageText);
        const contentState = ContentState.createFromBlockArray(blocksFromHTML);
        editorState = EditorState.push(this.state.editorState, contentState);
      } else {
        editorState = EditorState.createEmpty();
      }
      this.setState({editorState}, function () {
        this.focus();
      });
    }
```

**Comment by comatory** [comment]
If anybody is interested, I present my solution [here](https://codesandbox.io/s/2oyjo0m510). My editor clears its state when you lose focus. I also had to make sure that when you hit "edit" button again, the cursor is moved to the back.

**Comment by adrianmcli** [comment]
You may be using an older version of DraftJS Plugins? Note the FAQ question that was removed in this commit: https://github.com/draft-js-plugins/draft-js-plugins/commit/97f242ca04262e8b5ecc4a26d17ff1b8bb918b0b

**Comment by comatory** [comment]
@adrianmcli I think I'm using the latest versions:

```
    "draft-js": "^0.10.4",
    "draft-js-linkify-plugin": "^2.0.1",
    "draft-js-mention-plugin": "^2.0.1",
    "draft-js-plugins-editor": "^2.0.3",
```

I'm on React 15.4.2 though.

**Comment by comatory** [comment]
@adrianmcli I see that you are a contributor. First of all thanks for awesome work! It's very much appreciated.

It seems that the documentation does not say anything and clearing editor state should just work but there seems to be cases and people (like me) where it does not work (see https://github.com/draft-js-plugins/draft-js-plugins/issues/982).

Just wanted to bring this to your attention. Hopefully this will get fixed otherwise I really like the plugins, I dig the API for styling, I was able to customize everything and the stuff I didn't know I just looked at source code 👍
